hammerblade/torch/kernel/kernel_addmm.cpp

```diff
@@ -1,29 +1,183 @@
 //====================================================================
 // addmm kernel
-// 03/09/2020 YOUR NAME HERE Lin Cheng (YOU EMAIL HERE, lc873@cornell.edu)
+// 03/09/2020 Kexin Zheng, Lin Cheng (kz73@cornell.edu, lc873@cornell.edu)
 //====================================================================
 
 #include <kernel_common.hpp>
+#include <cmath>
+#define BLOCK_DIM 10 // sqrt(4KB/4 byte/4 data matrix) = 15 max 
 
 extern "C" {
 
   __attribute__ ((noinline))  int tensorlib_addmm(
-          bsg_tensor_t* self,
-          bsg_tensor_t* mat1,
-          bsg_tensor_t* mat2,
-          float* beta,
-          float* alpha) {
+          bsg_tensor_t* _result,
+          bsg_tensor_t* _self,
+          bsg_tensor_t* _mat1,
+          bsg_tensor_t* _mat2,
+          float* _beta,
+          float* _alpha) {
     // TODO: Convert uint32_t pointers to correct types
+    auto self = BSGTensor<float>(_self);
+    auto mat1 = BSGTensor<float>(_mat1);
+    auto mat2 = BSGTensor<float>(_mat2);
+    auto result = BSGTensor<float>(_result);
+    float beta = *_beta;
+    float alpha = *_alpha;
+
     // Start profiling
     bsg_cuda_print_stat_kernel_start();
+
     // TODO: Implement addmm
+
+    // v2: single tile, use blocking
+    int r1 = mat1.dim(0);
+    int c1 = mat1.dim(1);
+    int r2 = mat2.dim(0);
+    int c2 = mat2.dim(1);
+    assert(c1 == r2);
+
+    bsg_printf("r1 %d c1 %d r2 %d c2 %d\n", r1, c1, r2, c2);
+    bsg_printf("BLOCK_DIM %d\n", BLOCK_DIM);
+
+    // calculate number of row and col blocks in each matrix
+    int m1_num_blk_per_row = std::ceil((float)r1/(float)BLOCK_DIM); // how many blocks in m1 per row
+    int m1_num_blk_per_col = std::ceil((float)c1/(float)BLOCK_DIM); // how many blocks in m1 per col
+    int m2_num_blk_per_row = std::ceil((float)r2/(float)BLOCK_DIM); // how many blocks in m2 per row
+    int m2_num_blk_per_col = std::ceil((float)c2/(float)BLOCK_DIM); // how many blocks in m2 per col
```

Comment:
In terms of the optimizations, we'd rather prefer to have parallel naive implementation than high-performant single tile implementations. It could be more beneficial to put efforts in parallelization of your implementation first...

Moreover, classical matrix multiplication optimization might not work well for HB because of it's unique memory system.

Reply:
Hmmm ... blocking is critical for both the single tile and the parallel implementation. Also keep in mind that we are just getting started. And @kzheng9 and I agreed that we would start by focusing on a detailed data-drive optimization using blocking with a single tile and make sure we really understand that before moving on to parallelizing it (which should be relatively straight-forward once we have blocking working) ... and also keep in mind that we are well aware of the "unique memory system" of HB :)

Reply:
@kzheng9 avoid using floating point arithmetic to do this kind of block range calculation ... see if you can use integer arithmetic at all costs otherwise it just adds overhead

Reply:
Got it! I have modified my code so it only does integer arithmetic.